### PR TITLE
refactor: traitify column encodings in Read Buffer

### PR DESCRIPTION
**Pull request by e-dard**
## TLDR

This PR contributes to the current initiative of Query Performance by making it easier to add more encodings to the Read Buffer, which in turn will lead to a smaller in-memory footprint for table data and faster query execution.

More concretely, this PR means that the amount of code to add new encodings will be very small, generally just involving implementing a new _transcoder_ that defines how a value transforms from a logical type to a physical type.

This should mean that as I add things like:

 - RLE on integer columns;
 - converting natural floats into integer encodings;
 - frame of reference encoding;
 - specific time-series encoding for timestamps.

The amount of code change to the Read Buffer should be minimised.

There is a lot of code-churn here but the bottom line is this:

 - no behaviour change;
 - **~1,600** lines of code added but **~3,100** lines removed, so it's a net reduction in code.

A big thanks to @domodwyer who iterated on a prototype of this with me. I have asked him to review this as he will be most familiar with the design.

## Rationale

I want to add more encodings to the Read Buffer focussing on better compressing scalar data. This will allow IOx to store more data in memory for a given memory footprint. Compressing scalar values can also improve query execution performance because the compressed data can be operated on directly.

The current design of the encodings in the Read Buffer is that everything is specified concretely at compile time. The physical encoders (the objects responsible for storing data in memory and making it available for query) are generic over some _physical types_. However, they know nothing of:

(1) what the logical type of values they store are.
(2) how the values they store may need to be transformed when being emitted.

The information for `(1)` and `(2)` is stored in a higher level "logical type encoding", i.e., `IntegerEncoding` and `FloatEncoding`. These `enum`s store variants that describe all possible combinations of physical type, logical type and any transformation between them. For example:

```rust
enum IntegerEncoding{
  I64U8(Encoder<u8>),
  I64U16(Encoder<u16>),
  I64U32(Encoder<u32>),
  ...
  ...
}
```

Initially this was fine because there was only one transformation between logical and physical types (byte trimming) and it only applied to integers.

However, if I want to add more then I need to start incorporating the type of transformation into these `enum` variants. Plus, I want to start treating some floats as integers then I need to add many more variants there. 

It's clear that whilst this would work fine, it would be a lot of code that basically is just doing the job that the compiler can do for us!

## Approach

This PR adds a couple of traits to the Read Buffer `encoding` module that allow us to specify a couple of important shared behaviours. We can then implement these traits and _compose_ the implementations to get more combinations of columnar encoding without adding huge amounts of duplicate code.

#### Transcoder

```rust
pub trait Transcoder<P, L> {
    fn encode(&self, _: L) -> P;
    fn decode(&self, _: P) -> L;
}
```

A `Transcoder` describes how a scalar value of logical type `L` can be encoded as a scalar value of physical type `P`, and the inverse. The Read Buffer receives data in various logical types, e.g.., `i64`, `u64`, `f64`. An implementation of a `Transcoder` describes what we do to that data to get it specified as physically typed data `P`, ready to be stored in an encoding.

Currently in the Read Buffer we can already do byte trimming on integers. Byte trimming just means: if you have an input of  `Vec<20_i64, 30, 200>` (logical type `L = i64`) then you can transform that data into `Vec<20_u8, 30, 200>` (physical type `P = u8`) before you then encode it in a specific encoding. This transcoding is basically baked into the current encodings and the higher level logical encoding enums (e.g., `I64U8(...)`). What I have done in this PR is to remove that concept from being a part of the actual encodings and abstracted encapsulated it in an implementation of a `Transcoder`. It looks like this:

```rust 
pub struct ByteTrimmer {}

impl<P, L> Transcoder<P, L> for ByteTrimmer
where
    L: From<P>,
    P: TryFrom<L>,
    <P as TryFrom<L>>::Error: Debug,
{
    fn encode(&self, v: L) -> P {
        P::try_from(v).unwrap() // note: caller's responsible for ensuring this doesn't error.
    }

    fn decode(&self, v: P) -> L {
        L::from(v)
    }
}
```

As you might imagine, I should be able to  implement new encodings as transcoders in the future.

#### Scalar Encoding

The next trait is a `ScalarEncoding<L>` (generic over some logical type `L`), which describes the behaviour that an encoding has to provide. Effectively the API that can be used to query the encoded data. I won't list the trait here but will point it out in the PR code. The main reason for making this a trait is that it means we can really reduce the amount of code in the logical type encoders (`IntegerEncoding` and `FloatEncoding`) _and_ it means that when we change the transcoders that they use we don't need any more concrete enum variants. For example, before we had an `IntegerEncoding` with ~23 enum variants. If we added more transcodings then these variants would increase. Now though we just define `IntegerEncoding` like this:

```rust 
pub enum IntegerEncoding {
    // (encoding, name_of_encoding)
    I64(Box<dyn ScalarEncoding<i64>>, String),
    U64(Box<dyn ScalarEncoding<u64>>, String),
}
```

So each variant is an integer logical type that contains an implementation of a `ScalarEncoding<L>`. We can build _different_ `ScalarEncoding` at runtime by inspecting the contents of data that's given to the `IntegerEncoding` constructors (and the same for `FloatEncoding` naturally). 

**Note** I decided to move the trait as high as possible for performance reasons. I did not want to have to be doing a `vtable` lookup to decode every value in a column for example, so to make that work we push down a concrete `Transcoder` all the way down to each encoding.

Constructing an encoding looks like:

```rust 
let transcoder = ByteTrimmer{} // our chosen encoder that does byte trimming
let arr = arr // encode the input data (`Vec<L / u64>`)
            .iter()
            .map::<u8, _>(|v| transcoder.encode(*v)) // u64 -> u8
            .collect::<Vec<u8>>();
               
// give the encoder the encoded data (now in `Vec<P / u8>`) and transcoder
// so it knows how to decode values back to `L / u64`.
let enc = Box::new(Fixed::<u8, u64, _>::new(arr, transcoder));
```

To encode the same logically types column with  a different transcoder is just a case of providing a different `transcoder` at runtime.

## Next Steps

Once this PR is in I can start adding new transcoders to the Read Buffer with less code churn each time. I plan to add:

 - [ ] RLE integers
 - [ ] store floats as integers
 - [ ] frame of reference encoding
 - [ ] timestamp specific

### CODE REVIEW

File: read_buffer/benches/plain.rs [hunk 1]
```diff
@@ -1,11 +1,12 @@
+use rand::prelude::*;
```
Comment (e-dard) [inline read_buffer/benches/plain.rs:1]: Just getting these benchmarks to compile. Not much of interest here.

File: read_buffer/src/chunk.rs [hunk 2]
```diff
@@ -977,78 +977,90 @@ mod test {
...
-            r#"read_buffer_column_values{db="mydb",encoding="BT_U32",log_data_type="i64",null="false"} 9"#,
-            r#"read_buffer_column_values{db="mydb",encoding="BT_U32",log_data_type="i64",null="true"} 0"#,
-            r#"read_buffer_column_values{db="mydb",encoding="None",log_data_type="bool",null="false"} 9"#,
-            r#"read_buffer_column_values{db="mydb",encoding="None",log_data_type="bool",null="true"} 0"#,
-            r#"read_buffer_column_values{db="mydb",encoding="None",log_data_type="f64",null="false"} 15"#,
-            r#"read_buffer_column_values{db="mydb",encoding="None",log_data_type="f64",null="true"} 3"#,
-            r#"read_buffer_column_values{db="mydb",encoding="RLE",log_data_type="string",null="false"} 9"#,
-            r#"read_buffer_column_values{db="mydb",encoding="RLE",log_data_type="string",null="true"} 0"#,
-            "",
+                "# HELP read_buffer_column_bytes The number of bytes used by all columns in the Read Buffer",
```
Comment (e-dard) [inline read_buffer/src/chunk.rs:980]: Sizes of encodings changed slightly because of the introduction of a `Transcoder`. Plus I chaned the label names of the encodings slightly.

File: read_buffer/src/column/encoding/scalar.rs [hunk 3]
```diff
@@ -1,379 +1,91 @@
...
-    T: Debug + PartialOrd + Copy,
-    A: ArrowNumericType,
-{
-    Fixed(Fixed<T>),
-    FixedNullable(FixedNull<A>),
-    RLE(RLE<T>),
-}
+/// `ScalarEncoding` describes the behaviour of a columnar encoding for scalar
+/// values.
+pub trait ScalarEncoding<L>: Debug + Display + Send + Sync {
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar.rs:17]: This is the main trait that all scalar encodings have to implement.

File: read_buffer/src/column/encoding/scalar/fixed.rs [hunk 4]
```diff
@@ -1,51 +1,51 @@
...
-/// stored contiguous in a backing vector. Fixed encodings do not support NULL
+/// stored contiguously in a backing vector. Fixed encodings do not support NULL
 /// values, so are suitable for columns known to not have NULL values that we
 /// want to aggregate over.
 ///
-/// For a given logical datatype `U`, `Fixed` encodings can store values with
-/// a different datatype `T`, where `size_of::<T>() <= size_of::<U>()`.
-pub struct Fixed<T>
+/// Types are: Physical, Logical, Transcoder
+pub struct Fixed<P, L, T>
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/fixed.rs:29]: The only real change to the encoding is that now it's generic over a physical type, logical type and an implementation of a transcoder.

The transcoder is pushed into the encoding so that we don't need to be doing `vtable` lookups every time we want to decode a value.

I had to add a `PhantomData<L>` because the `T` needs to be generic over a `P` and `L`.

File: read_buffer/src/column/encoding/scalar/fixed.rs [hunk 5]
```diff
@@ -58,230 +58,16 @@ where
...
+impl<P, L, T> Fixed<P, L, T>
 where
-    T: std::fmt::Debug + PartialOrd + Copy,
+    P: Copy + Debug + PartialOrd,
+    T: Transcoder<P, L>,
 {
-    /// The name of this encoding.
-    pub fn name(&self) -> &'static str {
-        ENCODING_NAME
-    }
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/fixed.rs:68]: All of this has just moved in the file because it's now part of an `impl ScalarEncoding for Fixed` block.

File: read_buffer/src/column/encoding/scalar/fixed.rs [hunk 6]
```diff
@@ -484,276 +238,282 @@ where
     }
 }
 
-impl<T> iter::FromIterator<T> for Fixed<T>
+impl<P, L, T> ScalarEncoding<L> for Fixed<P, L, T>
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/fixed.rs:241]: Here is where the encoding implements `ScalarEncoding`.

File: read_buffer/src/column/encoding/scalar/fixed.rs [hunk 7]
```diff
@@ -484,276 +238,282 @@ where
...
+
+    fn has_any_non_null_value(&self) -> bool {
+        self.num_rows() > 0 // this encoding never contains NULL values
+    }
+
+    fn has_non_null_value(&self, _: &[u32]) -> bool {
+        self.num_rows() > 0 // this encoding never contains NULL values
+    }
+
+    fn value(&self, row_id: u32) -> Option<L> {
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/fixed.rs:276]: Here is an example of decoding a value to it's logical type `L` before returning it.

File: read_buffer/src/column/encoding/scalar/fixed.rs [hunk 8]
```diff
@@ -484,276 +238,282 @@ where
...
-    };
-}
+        }
+
+        // convert to logical type
+        Some(self.transcoder.decode(max))
+    }
+
+    fn row_ids_filter(&self, value: L, op: &cmp::Operator, dst: RowIDs) -> RowIDs {
+        let value = self.transcoder.encode(value);
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/fixed.rs:373]: Here is an example of encoding a predicate literal of type `L` into the stored physical type `P` so we can look for row IDs satisfying the filter predicate.

File: read_buffer/src/column/encoding/scalar/fixed.rs [hunk 9]
```diff
@@ -484,276 +238,282 @@ where
...
-
-    #[test]
-    fn first_row_id_eq_value() {
-        let v: Fixed<i64> = Fixed {
-            values: vec![22, 33, 18],
-        };
-
-        assert_eq!(v.first_row_id_eq_value(33), Some(1));
-        assert_eq!(v.first_row_id_eq_value(100), None);
+    // Helper function to create a new FixedNull encoding using a mock transcoder
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/fixed.rs:419]:
```suggestion
    // Helper function to create a new Fixed encoding using a mock transcoder
```

will change.

File: read_buffer/src/column/encoding/scalar/fixed.rs [hunk 10]
```diff
@@ -484,276 +238,282 @@ where
...
+    // that will allow tests to track calls to encode/decode.
+    fn new_encoding(
+        values: Vec<i64>,
+    ) -> (Fixed<i64, i64, Arc<MockTranscoder>>, Arc<MockTranscoder>) {
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/fixed.rs:423]: You can get an idea of how `MockTranscoder` works in the `transcoders.rs` file.

File: read_buffer/src/column/encoding/scalar/fixed.rs [hunk 11]
```diff
@@ -484,276 +238,282 @@ where
...
-        // 3]).is_nan());
+        assert_eq!(v.max(&[0, 1, 2, 3, 4]), Some(4));
+        assert_eq!(transcoder.decodings(), 1); // only max value decoded
     }
 
     #[test]
     fn ord_from_op() {
         assert_eq!(
-            Fixed::<i64>::ord_from_op(&cmp::Operator::LT),
+            Fixed::<i64, i64, NoOpTranscoder>::ord_from_op(&cmp::Operator::LT),
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/fixed.rs:488]: A `NoOpTranscoder` does what it says on the tin - it just spits back the same value you encode/decode.

File: read_buffer/src/column/encoding/scalar/fixed_null.rs [hunk 12]
```diff
@@ -1,51 +1,51 @@
...
+use super::transcoders::Transcoder;
+use super::ScalarEncoding;
+use crate::column::{cmp, RowIDs};
+
 pub const ENCODING_NAME: &str = "FIXEDN";
 
 #[derive(Debug, PartialEq)]
-pub struct FixedNull<T>
+/// Types are: Physical, Logical, Transcoder
+pub struct FixedNull<P, L, T>
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/fixed_null.rs:28]: If you have reviewed the `Fixed` encoding you're basically going to see something very similar here. The real difference is just that this is backed by an Arrow array, which supports NULL values but has a larger footprint than `Fixed`.

File: read_buffer/src/column/encoding/scalar/rle.rs [hunk 13]
```diff
@@ -1,22 +1,24 @@
...
 #[allow(clippy::upper_case_acronyms)] // this looks weird as `Rle`
 /// An RLE encoding is one where identical "runs" of values in the column are
 /// stored as a tuple: `(run_length, value)`, where `run_length` indicates the
 /// number of times the value is to be repeated.
-#[derive(Debug, Default, PartialEq)]
-pub struct RLE<T>
+///
+/// Types are: Physical, Logcial, Transcoder
+#[derive(Debug)]
+pub struct RLE<P, L, T>
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/rle.rs:19]: As with the other encodings I define `RLE` in terms of a logical type, physical type and a transcoder.

File: read_buffer/src/column/encoding/scalar/transcoders.rs [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    fn decode(&self, _: P) -> L;
+}
+
+/// A No-op transcoder - useful when you can't do any useful transcoding, e.g.,
+/// because you have a very high entropy 64-bit column.
+///
+/// `NoOpTranscoder` implements `Transcoder` in terms of `T` -> `T` and just
+/// returns the provided argument.
+#[derive(Debug)]
+pub struct NoOpTranscoder {}
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/transcoders.rs:26]: Useful when you can't do anything with a column of data - just spits back whatever `T` you encode or decode.

File: read_buffer/src/column/encoding/scalar/transcoders.rs [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+#[derive(Debug)]
+pub struct ByteTrimmer {}
+impl<P, L> Transcoder<P, L> for ByteTrimmer
+where
+    L: From<P>,
+    P: TryFrom<L>,
+    <P as TryFrom<L>>::Error: std::fmt::Debug,
+{
+    fn encode(&self, v: L) -> P {
+        P::try_from(v).unwrap()
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/transcoders.rs:63]: The design of the Read Buffer is such that higher level encoders have to determine that all the data can be byte trimmed correctly.

File: read_buffer/src/column/encoding/scalar/transcoders.rs [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+//        to logical values and then optionally apply a byte trimming to the
+//        result.
+
+#[cfg(test)]
+use std::{sync::atomic::AtomicUsize, sync::atomic::Ordering, sync::Arc};
+#[cfg(test)]
+/// A mock implementation of Transcoder that tracks calls to encode and decode.
+/// This is useful for testing encoder implementations.
+#[derive(Debug)]
+pub struct MockTranscoder {
```
Comment (e-dard) [inline read_buffer/src/column/encoding/scalar/transcoders.rs:93]: This is useful for testing that the transcoder is wired in correctly. I had to just atomic primitives for this because the Read Buffer chunks (and their contents including the transcoders) need to be `Sync+Send`.

File: read_buffer/src/column/float.rs [hunk 17]
```diff
@@ -11,24 +13,19 @@ use crate::column::{RowIDs, Scalar, Value, Values};
...
 }
 
 impl FloatEncoding {
     /// The total size in bytes of to store columnar data in memory.
     pub fn size(&self) -> usize {
         match self {
-            Self::Fixed64(enc) => enc.size(),
-            Self::FixedNull64(enc) => enc.size(),
-            Self::RLE64(enc) => enc.size(),
+            Self::F64(enc) => enc.size(),
```
Comment (e-dard) [inline read_buffer/src/column/float.rs:28]: Mainly code deletion! I left it as an `enum` for similarity to `IntegerEncoding`.

File: read_buffer/src/column/float.rs [hunk 18]
```diff
@@ -11,24 +13,19 @@ use crate::column::{RowIDs, Scalar, Value, Values};
...
 pub enum FloatEncoding {
-    // A fixed-width "no compression" vector of non-nullable values
-    Fixed64(Fixed<f64>),
-
-    // A fixed-width "no compression" vector of nullable values (as Arrow array)
-    FixedNull64(FixedNull<arrow::datatypes::Float64Type>),
-
-    // A RLE compressed encoding of nullable values.
-    RLE64(RLE<f64>),
+    F64(Box<dyn ScalarEncoding<f64>>),
```
Comment (e-dard) [inline read_buffer/src/column/float.rs:21]: Mainly code deletion! I left it as an `enum` for similarity to `IntegerEncoding`.

File: read_buffer/src/column/float.rs [hunk 19]
```diff
@@ -332,43 +284,32 @@ impl From<arrow::array::Float64Array> for FloatEncoding {
         let base_size = arr.len() * size_of::<f64>();
         let rle_size = rle_rows_opt(arr.iter()) * size_of::<(u32, Option<f64>)>(); // size of a run length
         if (base_size as f64 - rle_size as f64) / base_size as f64 >= MIN_RLE_SIZE_REDUCTION {
-            return Self::RLE64(RLE::from_iter(&arr));
+            let enc = Box::new(RLE::new_from_iter_opt(
```
Comment (e-dard) [inline read_buffer/src/column/float.rs:287]: Here is where I can add logic in the future to choose different encodings. For example we could potentially convert some natural number float columns into integer columns and then byte trim them, and then RLE them. It would just be a case of adding logic here and providing a transcoder implementation that does that.

File: read_buffer/src/column/float.rs [hunk 20]
```diff
@@ -332,43 +284,32 @@ impl From<arrow::array::Float64Array> for FloatEncoding {
...
+                arr.iter(),
+                NoOpTranscoder {}, // No transcoding of values (store as physical type f64)
```
Comment (e-dard) [inline read_buffer/src/column/float.rs:289]: To keep this PR as purely a refactor I kept the existing behaviour, which is that floats do not get transformed before they are stored in an encoding.

File: read_buffer/src/column/integer.rs [hunk 21]
```diff
@@ -1,209 +1,52 @@
...
-            (Self::U64U8N(a), Self::U64U8N(b)) => {
-                let a = a.all_values::<u64>(vec![]);
-                let b = b.all_values(vec![]);
-                a == b
-            }
-            (_, _) => false,
-        }
-    }
+    // (encoding, name_of_encoding)
+    I64(Box<dyn ScalarEncoding<i64>>, String),
```
Comment (e-dard) [inline read_buffer/src/column/integer.rs:21]: This is better :-)

File: read_buffer/src/column/integer.rs [hunk 22]
```diff
@@ -960,31 +267,74 @@ impl From<&[i64]> for IntegerEncoding {
 
         // This match is carefully ordered. It prioritises smaller physical
         // datatypes that can safely represent the provided logical data
+        let transcoder = ByteTrimmer {};
```
Comment (e-dard) [inline read_buffer/src/column/integer.rs:270]: Right now the `IntegerEncoding` should perform byte trimming if possible. This is the logic to do that, which isn't any different than before. The difference is that we initialise different implementations of `ScalarEncoding` and then store the boxed trait on the `IntegerEncoding`.

File: read_buffer/src/column/integer.rs [hunk 23]
```diff
@@ -960,31 +267,74 @@ impl From<&[i64]> for IntegerEncoding {
...
+                    .map(|v| transcoder.encode(*v))
+                    .collect::<Vec<i32>>();
+                let enc = Box::new(Fixed::<i32, i64, _>::new(arr, transcoder));
+                let name = enc.name();
+                Self::I64(enc, format!("BT_I32-{}", name))
             }
             // otherwise, encode with the same physical type (i64)
-            (_, _) => Self::I64I64(Fixed::<i64>::from(arr)),
+            (_, _) => {
+                let enc = Box::new(Fixed::<i64, i64, _>::new(arr.to_vec(), NoOpTranscoder {}));
```
Comment (e-dard) [inline read_buffer/src/column/integer.rs:334]: We can't trim the values in this data so we can just use a `NoOpTranscoder {}`.

File: read_buffer/src/column/encoding/scalar/fixed.rs (hunk 4 above)
Comment (tustvold) [reply read_buffer/src/column/encoding/scalar/fixed.rs:29]: Why not make P and L associated types of the transcoder?

File: read_buffer/src/column/encoding/scalar/fixed.rs [hunk 24]
```diff
@@ -484,276 +238,282 @@ where
...
+        self.values.len() as u32
+    }
+
+    /// Encoded data size including `Self` - an "accurate" estimation.
+    fn size(&self) -> usize {
+        size_of::<Self>() + (size_of::<P>() * self.values.len())
+    }
+
+    fn size_raw(&self, _: bool) -> usize {
+        size_of::<Vec<L>>() + (size_of::<L>() * self.values.len())
```
[the remark this answers is not in the document]
Comment (e-dard) [reply read_buffer/src/column/encoding/scalar/fixed.rs:261]: That would be the size of a vec of `P` but we want a size of `L` (as if no encoding had been done to the logical values).

File: read_buffer/src/column/encoding/scalar/fixed.rs [hunk 25]
```diff
@@ -484,276 +238,282 @@ where
...
+            dst.push(self.transcoder.decode(self.values[chunks[0] as usize]));
+            dst.push(self.transcoder.decode(self.values[chunks[1] as usize]));
+            dst.push(self.transcoder.decode(self.values[chunks[2] as usize]));
+            dst.push(self.transcoder.decode(self.values[chunks[3] as usize]));
+        }
+
+        let rem = row_ids.len() % 4;
+        for &i in &row_ids[row_ids.len() - rem..row_ids.len()] {
+            dst.push(self.transcoder.decode(self.values[i as usize]));
         }
```
[the remark this answers is not in the document]
Comment (e-dard) [reply read_buffer/src/column/encoding/scalar/fixed.rs:295]: I spent a while trying all the different ways to do this last year and the chunking one came out on top. Happy to benchmark  something else but I think it's best left for another PR.

File: read_buffer/src/column/encoding/scalar.rs [hunk 26]
```diff
@@ -1,379 +1,91 @@
...
-        match self {
-            Self::Fixed(_) => !row_ids.is_empty(), // all rows will be non-null
-            Self::FixedNullable(c) => c.has_non_null_value(row_ids),
-            Self::RLE(c) => c.has_non_null_value(row_ids),
-        }
-    }
+    /// Returns the logical values found at the provided ordinal offsets.
+    /// Implementations that can't contain NULL values have the option to return
+    /// `Either::Left`.
+    fn values(&self, row_ids: &[u32]) -> Either<Vec<L>, Vec<Option<L>>>;
```
[the remark this answers is not in the document]
Comment (e-dard) [reply read_buffer/src/column/encoding/scalar.rs:49]: haha so they did accept a vec but it's problematic because of `Either`. I was going to figure it out in another PR. You're spot on 👍

File: read_buffer/src/chunk.rs [hunk 27]
```diff
@@ -675,7 +675,7 @@ impl ChunkMetrics {
     fn update_column_storage_statistics(&mut self, statistics: &[Statistics]) {
         for stat in statistics {
             let labels = &[
-                KeyValue::new("encoding", stat.enc_type),
+                KeyValue::new("encoding", stat.enc_type.clone()),
```
[the remark this answers is not in the document]
Comment (e-dard) [reply read_buffer/src/chunk.rs:678]: Thanks. I will address this in a follow-up PR. Great explanation!

[the remark this answers is not in the document]
Comment (alamb) [reply read_buffer/src/chunk.rs:678]: I have always found the `Cow` name amusing 🐮  to type but misleading because almost all the usecases for `Cow` I have stumbled across are actually for the abstraction that @domodwyer  is describing

File: read_buffer/src/column/encoding/scalar.rs [hunk 28]
```diff
@@ -1,379 +1,91 @@
...
-        format!(
-            "SCALAR_{}",
-            match &self {
-                Self::Fixed(c) => c.name(),
-                Self::FixedNullable(c) => c.name(),
-                Self::RLE(c) => c.name(),
-            }
-        )
-    }
+    /// The total size in bytes to store encoded data in memory.
```
Comment (alamb) [inline read_buffer/src/column/encoding/scalar.rs:21]: Is this (intended) to be "allocated" or "used" size?

File: read_buffer/src/column/encoding/scalar.rs (hunk 3 above)
Comment (alamb) [reply read_buffer/src/column/encoding/scalar.rs:17]: This is a cool design. 👍

File: read_buffer/src/column/encoding/scalar/transcoders.rs (hunk 15 above)
Comment (alamb) [reply read_buffer/src/column/encoding/scalar/transcoders.rs:63]: that makes sense -- eventually it would be cool to avoid the runtime overhead of `unwrap()` (by using some infallible trait) but for another PR if it turns out to be too much overhead

**Review by alamb — APPROVED** [review]
I went over the high level design of this PR and I think it is really cool / clever. 👍  Very cool

The few places I looked at the code in detail  looked cool too.

File: read_buffer/src/column/encoding/scalar.rs (hunk 28 above)
Comment (e-dard) [reply read_buffer/src/column/encoding/scalar.rs:21]: Used size because I didn't want to change behaviour in the PR. I'm going to change to capacity size in a follow on (I don't expect it to make a huge difference in reported size)

Comment (alamb) [reply read_buffer/src/column/encoding/scalar.rs:21]: makes sense
